stan/math/opencl/kernels/multiply_transpose.hpp

```diff
@@ -63,23 +63,24 @@ static const char* multiply_transpose_kernel_code = STRINGIFY(
                    [thread_block_row]
                 = A[(j + w * THREAD_BLOCK_SIZE_COL) + tiled_i * M];
           }
-        }
-        // wait till all tile values are loaded to the local memory
-        barrier(CLK_LOCAL_MEM_FENCE);
-        // multiply the tile products
-        for (int block_ind = 0; block_ind < THREAD_BLOCK_SIZE; block_ind++) {
-          // each thread multiplies WORK_PER_THREAD values
-          for (int w = 0; w < WORK_PER_THREAD; w++) {
-            if (jMin <= iMax) {
-              if ((j + w * THREAD_BLOCK_SIZE_COL) <= i) {
-                acc[w] += A_local[block_ind][thread_block_row]
-                          * B_local[thread_block_col
-                                    + w * THREAD_BLOCK_SIZE_COL][block_ind];
+
+          // wait till all tile values are loaded to the local memory
+          barrier(CLK_LOCAL_MEM_FENCE);
+          // multiply the tile products
+          for (int block_ind = 0; block_ind < THREAD_BLOCK_SIZE; block_ind++) {
+            // each thread multiplies WORK_PER_THREAD values
+            for (int w = 0; w < WORK_PER_THREAD; w++) {
+              if (jMin <= iMax) {
+                if ((j + w * THREAD_BLOCK_SIZE_COL) <= i) {
+                  acc[w] += A_local[block_ind][thread_block_row]
+                            * B_local[thread_block_col
+                                      + w * THREAD_BLOCK_SIZE_COL][block_ind];
+                }
               }
             }
           }
+          barrier(CLK_LOCAL_MEM_FENCE);
         }
-        barrier(CLK_LOCAL_MEM_FENCE);
       }
```

Comment:
Hah, this misplaced `}` seems like a pretty major bug. If you haven't yet, could you write a regression test that fails on the old version? Thanks :D

Reply:
Luckily in this instance its just a performance issue, not a bug. The values `jMin` and `iMax` determine if the thread actually does anything or it could just skip the entire calculation. In the previous location all the threads were syncing all the time, even those that could skip everything. That slowed things down a bit, but it did not corrupt the resulting matrix.

![Screenshot from 2019-03-13 17-35-15](https://user-images.githubusercontent.com/28476796/54297218-b97f4180-45b6-11e9-82f5-085fd1d18aa7.png)

`if (jMin <= iMax) {` eliminates threads that are responsible for the red parts of the resulting matrix. Now that I had to think of this again, this check could be moved to the outmost level as these threads actually dont need to work at all.

All threads responsible for the grey tiles need to at least copy the data to the shared memory (to "help" the threads in the tile that actually need to multiply). The `if ((j + w * THREAD_BLOCK_SIZE_COL) <= i) {` (line 74 & 88) check eliminates the threads above the diagonal in grey tiles from multiplication and writing to the result matrix.